### PR TITLE
Httpclient improvements

### PR DESCRIPTION
**Pull request by planetis-m**
Rebase old pr #10670 by @LemonBoy.

### CODE REVIEW

**Comment by planetis-m** [comment]
I think it's also worth redoing this [commit](https://github.com/nim-lang/Nim/pull/10668/commits/7ead616d4c02733bd9a728ffa03e00c0c1d9e69c) that uses HttpMethod instead of string.

**Comment by planetis-m** [comment]
Ready for review @Araq, @dom96

**Comment by planetis-m** [comment]
This needs a review @dom96 it's ready

File: lib/pure/httpclient.nim [hunk 1]
```diff
@@ -325,9 +325,12 @@ proc getDefaultSSL(): SslContext =
       result = defaultSslContext
       doAssert result != nil, "failure to initialize the SSL context"
 
-proc newProxy*(url: string, auth = ""): Proxy =
+proc newProxy*(url: Uri | string, auth = ""): Proxy =
```
Comment (dom96) [inline lib/pure/httpclient.nim]: Instead of doing this, overload the procs.

File: lib/pure/httpclient.nim [hunk 2]
```diff
@@ -457,29 +460,23 @@ proc sendFile(socket: Socket | AsyncSocket,
...
+    result.anchor = parsedLocation.anchor
+  else:
+    result = parsedLocation
+
+proc generateHeaders(requestUrl: Uri, httpMethod: HttpMethod, headers: HttpHeaders,
                      proxy: Proxy): string =
   # GET
-  let upperMethod = httpMethod.toUpperAscii()
-  result = upperMethod
+  result = $httpMethod
```
Comment (dom96) [inline lib/pure/httpclient.nim:480]: Why this change? Is the output the same? (uppercase?)

File: lib/pure/httpclient.nim [hunk 3]
```diff
@@ -1037,33 +1033,86 @@ proc request*(client: HttpClient | AsyncHttpClient, url: string,
   ##
   ## You need to make sure that the ``url`` doesn't contain any newline
   ## characters. Failing to do so will raise ``AssertionDefect``.
-  doAssert(not url.contains({'\c', '\L'}), "url shouldn't contain any newline characters")
+  ## **Deprecated since v1.5**: use HttpMethod enum instead; string parameter httpMethod is deprecated
```
Comment (dom96) [inline lib/pure/httpclient.nim:1041]:
```suggestion
  ## 
  ## **Deprecated since v1.5**: use HttpMethod enum instead; string parameter httpMethod is deprecated
```

File: lib/pure/httpclient.nim [hunk 4]
```diff
@@ -1037,33 +1033,86 @@ proc request*(client: HttpClient | AsyncHttpClient, url: string,
...
+  when url is string:
+    doAssert(not url.contains({'\c', '\L'}), "url shouldn't contain any newline characters")
```
Comment (dom96) [inline lib/pure/httpclient.nim:1043]: This seems like it should be thrown by `parseUri` below rather than handled here.

File: lib/pure/httpclient.nim [hunk 5]
```diff
@@ -1037,33 +1033,86 @@ proc request*(client: HttpClient | AsyncHttpClient, url: string,
...
+    of Http301, Http302, Http303:
+      # The method is changed to GET unless it is GET or HEAD (RFC2616)
+      if httpMethod notin {HttpGet, HttpHead}:
+        redirectMethod = HttpGet
+      else:
+        redirectMethod = httpMethod
+      # The body is stripped away
+      redirectBody = ""
+      # Delete any header value associated with the body
+      if headers != nil:
```
Comment (dom96) [inline lib/pure/httpclient.nim]:
```suggestion
      if not headers.isNil():
```

File: lib/pure/httpclient.nim [hunk 6]
```diff
@@ -1037,33 +1033,86 @@ proc request*(client: HttpClient | AsyncHttpClient, url: string,
...
+    # Check if the redirection is to the same domain or a sub-domain (foo.com
+    # -> sub.foo.com)
+    if redirectTo.hostname != lastURL.hostname and
+      not redirectTo.hostname.endsWith("." & lastURL.hostname):
+      # Perform some cleanup of the header values
+      if headers != nil:
+        # Delete the Host header
+        headers.del("Host")
+        # Do not send any sensitive info to a unknown host
+        headers.del("Authorization")
```
Comment (dom96) [inline lib/pure/httpclient.nim:1116]: Is this behaviour part of the HTTP spec?

File: tests/stdlib/thttpclient.nim [hunk 7]
```diff
@@ -39,7 +39,7 @@ proc makeIPv6HttpServer(hostname: string, port: Port,
 
 proc asyncTest() {.async.} =
   var client = newAsyncHttpClient()
-  var resp = await client.request("http://example.com/")
+  var resp = await client.request("http://example.com/", "GET")
```
Comment (dom96) [inline tests/stdlib/thttpclient.nim]: Our tests shouldn't have deprecations warnings ideally.

**Review by dom96 — APPROVED** [review]
Overall looks good. Please resolve my comments though.

Also, would be nice if we could get test coverage for this redirect logic.

File: lib/pure/httpclient.nim (hunk 2 above)
Comment (planetis-m) [reply lib/pure/httpclient.nim:480]: yes, variable `httpMethod` changed from `string` to `enum` and the `HttpMethod` enum has all uppercase string values (defined in httpcore.nim)

File: lib/pure/httpclient.nim (hunk 4 above)
Comment (planetis-m) [reply lib/pure/httpclient.nim:1043]: Agreed, should be handled by uri. Maybe #15028 has to be done differently or I can put the condition elsewere.

Comment (planetis-m) [reply lib/pure/httpclient.nim:1043]: Well I don't really understand the original PR intentions (nor the Uri spec)  should be done in a seperate PR on uri module, for now I leave it as is.

File: lib/pure/httpclient.nim (hunk 6 above)
Comment (planetis-m) [reply lib/pure/httpclient.nim:1116]: I don't think it's specified (in HTTP/1.0), but seems like a good security measure.

File: lib/pure/httpclient.nim [hunk 8]
```diff
@@ -1087,7 +1089,7 @@ proc request*(client: HttpClient | AsyncHttpClient, url: Uri | string,
       # The body is stripped away
       redirectBody = ""
       # Delete any header value associated with the body
-      if headers != nil:
+      if not headers.isNil():
```
Comment (xflywind) [inline lib/pure/httpclient.nim:1092]: Why use `isNil`? AFAIR  `!= nil` style may be preferred.

Comment (planetis-m) [reply lib/pure/httpclient.nim:1092]: :shrug:

Comment (xflywind) [reply lib/pure/httpclient.nim:1092]: https://discord.com/channels/371759389889003530/371759389889003532/696718991019999252

File: lib/pure/httpclient.nim (hunk 6 above)
Comment (dom96) [reply lib/pure/httpclient.nim:1116]: If it's not part of the spec then we shouldn't introduce it. HTTP is ultimately responsible for security in this context AFAIK and this may break existing applications.

Comment (planetis-m) [reply lib/pure/httpclient.nim:1116]: @dom96 Have you checked it's not? Im not the original author of this commit (as it can be seen from the git log)

File: lib/pure/httpclient.nim (hunk 4 above)
Comment (dom96) [reply lib/pure/httpclient.nim:1043]: Instead of guessing intentions let's do what we deem right. Otherwise we will end up with code that nobody understands except the original person who is no longer around to answer for it :)

File: lib/pure/httpclient.nim (hunk 6 above)
Comment (dom96) [reply lib/pure/httpclient.nim:1116]: Okay, I just looked into it. There is some evidence that it is done by other HTTP clients at least: https://stackoverflow.com/a/28671822/492186

File: lib/pure/httpclient.nim (hunk 4 above)
Comment (dom96) [reply lib/pure/httpclient.nim:1043]: Since this is the only thing I think needs to be changed though, can you create a separate PR for this? I will merge this.

**Review by dom96 — APPROVED** [review]

**Comment by planetis-m** [comment]
Thanks for merging this!
